lib/class-wp-theme-json-gutenberg.php

```diff
@@ -2855,6 +2866,11 @@ static function ( $pseudo_selector ) use ( $selector ) {
 			$block_rules .= static::to_ruleset( ":root :where($style_variation_selector)", $individual_style_variation_declarations );
 		}
 
+		// 7. Generate and append any custom CSS rules.
+		if ( isset( $node['css'] ) ) {
+			$block_rules .= $this->process_blocks_custom_css( $node['css'], $selector );
```

Comment:
I think this line allows adding custom CSS declarations to the root selector in theme.json

```json
	"styles": {
		"css": "background-color:purple;"
	},
```

I can see the background color on the frontend but not in the site editor

Whereas in trunk only this works:

```json
	"styles": {
		"css": "body{background-color:purple;}"
	},
```

Should Gutenberg allow both? Or is it a regression?

Reply:
Hmm that's a good question. This is how it was before #47396, but given that PR is over a year old it might be best to keep things as they were more recently. I'd expect to have to write complete rules in the top level custom CSS control and not necessarily for loose properties to be attached to the body element.

Reply:
Oooooh this is interesting; I think we've detected a bug in the existing behaviour. Testing in trunk, the loose properties are output outside any rule, so they don't apply because they're invalid, both in editor and front end:

<img width="409" alt="Screenshot 2024-06-11 at 3 43 26 PM" src="https://github.com/WordPress/gutenberg/assets/8096000/2780ce3b-7ef9-4eec-a3d8-ba49078a9927">

<img width="696" alt="Screenshot 2024-06-11 at 3 47 01 PM" src="https://github.com/WordPress/gutenberg/assets/8096000/5175b018-762d-4204-93d2-53f78377c072">

Whereas in this branch, those properties are nested under the root selector in the front end, though they're still invalid output in the editor. I think ideally we should fix things so they aren't output at all. I'll look into it!

Reply:
Ok I fixed things so that this branch behaves like trunk (no valid CSS output if only properties are added to the top level custom CSS control) but the loose properties are still being output (just like they are on trunk).

The problem with this is that any rule output straight after a loose property won't work. I can repro this in core trunk, and it's also a problem with the customizer additional CSS (which may not manifest in classic themes because additional CSS is enqueued after everything else, but on hybrid themes we're adding it before global styles custom CSS, so it can also break stuff)

We could maybe just check that there aren't any loose properties at the end of the custom CSS string? Checking the whole string is valid CSS might be overkill as site admins should be responsible for the validity of their CSS. We could also Do Nothing 😄 and leave things as they are.

If we do want to do anything though, it's probably better as a separate PR.

Reply:
> If we do want to do anything though, it's probably better as a separate PR.

For sure. Thanks for looking into it. 

> We could maybe just check that there aren't any loose properties at the end of the custom CSS string? 

Or maybe some light-weight parsing where the styles are added?

https://github.com/WordPress/gutenberg/blob/01a0e429483e85245112219bb66f12a65063014b/lib/class-wp-theme-json-gutenberg.php#L1361-L1361